examples/kitt/agent.py

```diff
@@ -0,0 +1,179 @@
+    audio_stream = await audio_stream_future
+
+    async def audio_stream_task():
+        print("Starting audio stream")
+        async for audio_frame_event in audio_stream:
+            # Ignore user input while the agent is speaking
+            if agent_speaking:
+                continue
+            stt_stream.push_frame(audio_frame_event.frame)
+
+    async def stt_stream_task():
+        nonlocal working_text, latest_llm_stream, latest_tts_stream
+        async for stt_event in stt_stream:
+            # We eagerly try to run inference to keep the latency as low as possible.
+            # If we get a new transcript, we update the working text, cancel in-flight inference,
+            # and run new inference.
+            if stt_event.type == agents.stt.SpeechEventType.FINAL_TRANSCRIPT:
+                delta = stt_event.alternatives[0].text
+                # Do nothing
+                if delta == "":
+                    continue
+                working_text += delta
+                print("Working text: ", working_text)
+                working_message = ChatMessage(role=ChatRole.USER, text=working_text)
+                chat_context = agents.llm.ChatContext(
+                    messages=committed_messages + [working_message]
+                )
+                # Cancel exisiting LLM inference
+                if latest_llm_stream is not None:
+                    await latest_llm_stream.aclose(wait=False)
+                latest_llm_stream = await llm.chat(history=chat_context)
+                # Cancel existing tts inference
+                await latest_tts_stream.aclose(wait=False)
+                # Pre-warm the tts
+                latest_tts_stream = tts.stream()
+                await llm_stream_queue.put(latest_llm_stream)
+
+    async def chat_stream_task():
+        while True:
+            llm_stream = await llm_stream_queue.get()
+            if llm_stream is None:
+                break
+            await tts_stream_queue.put(latest_tts_stream)
+            async for chunk in llm_stream:
+                print("Got chunk: ", chunk)
+                content = chunk.choices[0].delta.content
+                latest_tts_stream.push_text(content)
+            await latest_tts_stream.flush()
+
+    async def tts_stream_task():
+        nonlocal agent_speaking, working_text
+        while True:
+            tts_stream = await tts_stream_queue.get()
+            if tts_stream is None:
+                break
+            agent_speaking = True
+            committed_message = False
+            async for event in tts_stream:
+                if event.type == agents.tts.SynthesisEventType.AUDIO:
+                    # As soon as the agent speaks, we commit the working text
+                    # and the agent's speech to the chat.
+
+                    if not committed_message:
+                        committed_messages.append(
+                            ChatMessage(role=ChatRole.USER, text=working_text),
```

Comment:
this seems a bit dangerous to access working_text from outside of the current task. we don't really know the sequencing of things with this right? i.e. is it possible to have:

1. user speaking a paragraph (p1)
2. llm processed p1
3. tts streaming response to that
4. user spoke more (p2a)
5. we are still processing tts_stream, and p2a ends up getting added to the history in the wrong order.